SYCL/UserDefinedReductions/user_defined_reductions.cpp

```diff
@@ -0,0 +1,209 @@
+}
+
+struct custom_type {
+  static constexpr unsigned long long default_ull_value = 42;
+
+  constexpr custom_type() = default;
+  constexpr custom_type(int i, float f, unsigned long long ull)
+      : n(i, f), ull(ull) {}
+
+  custom_type_nested n;
+  unsigned long long ull = default_ull_value;
+};
+
+inline bool operator==(const custom_type &lhs, const custom_type &rhs) {
+  return lhs.n == rhs.n && lhs.ull == rhs.ull;
+}
+
+inline custom_type operator+(const custom_type &lhs, const custom_type &rhs) {
+  return custom_type(lhs.n.i + rhs.n.i, lhs.n.f + rhs.n.f, lhs.ull + rhs.ull);
+}
+
+struct custom_type_wo_default_ctor {
+  static constexpr unsigned long long default_ull_value = 42;
+
+  constexpr custom_type_wo_default_ctor() = delete;
+  constexpr custom_type_wo_default_ctor(int i, float f, unsigned long long ull)
+      : n(i, f), ull(ull) {}
+
+  custom_type_nested n;
+  unsigned long long ull = default_ull_value;
+};
+
+inline bool operator==(const custom_type_wo_default_ctor &lhs,
+                       const custom_type_wo_default_ctor &rhs) {
+  return lhs.n == rhs.n && lhs.ull == rhs.ull;
+}
+
+inline custom_type_wo_default_ctor
+operator+(const custom_type_wo_default_ctor &lhs,
+          const custom_type_wo_default_ctor &rhs) {
+  return custom_type_wo_default_ctor(lhs.n.i + rhs.n.i, lhs.n.f + rhs.n.f,
+                                     lhs.ull + rhs.ull);
+}
+
+template <typename T, std::size_t... Is>
+constexpr std::array<T, sizeof...(Is)> init_array(T value,
+                                                  std::index_sequence<Is...>) {
+  return {{(static_cast<void>(Is), value)...}};
+}
+
+using namespace sycl;
+
+template <typename InputContainer, typename OutputContainer,
+          class BinaryOperation>
+void test(queue q, InputContainer input, OutputContainer output,
+          BinaryOperation binary_op, size_t workgroup_size,
+          typename OutputContainer::value_type identity,
+          typename OutputContainer::value_type init) {
+  using InputT = typename InputContainer::value_type;
+  using OutputT = typename OutputContainer::value_type;
+  constexpr size_t N = input.size();
+  {
+    buffer<InputT> in_buf(input.data(), input.size());
+    buffer<OutputT> out_buf(output.data(), output.size());
+
+    q.submit([&](handler &cgh) {
+      accessor in{in_buf, cgh, sycl::read_only};
+      accessor out{out_buf, cgh, sycl::write_only, sycl::no_init};
+
+      size_t temp_memory_size = workgroup_size * sizeof(InputT);
+      auto scratch = sycl::local_accessor<std::byte, 1>(temp_memory_size, cgh);
+      cgh.parallel_for(
+          nd_range<1>(workgroup_size, workgroup_size), [=](nd_item<1> it) {
+            // Create a handle that associates the group with an allocation it
+            // can use
+            auto handle =
+                sycl::ext::oneapi::experimental::group_with_scratchpad(
+                    it.get_group(), sycl::span(&scratch[0], temp_memory_size));
```

Comment:
I'm still not sure if it has non-pointer iterators underneath. Are you sure that scenario is tested now?